### ISSUE TITLE
(potential?) bug: AVR - Incorrect f64 handling

### ISSUE THREAD

**Issue by onkoe** [rust-lang/rust#118079]
Hi there! I encountered a weird situation when writing a device driver using `embedded-hal`. Despite getting correct temperature readings in byte form from my thermocouple converter, all readings were incorrect after casting them to `f64` floats and applying some fundamental transformations.

After much confusion, I tried using `f32` instead and found it worked perfectly!

However, I knew the transformations should be correct either way: I tested this exact setup and formula using Linux and the `spidev` crate. 

That led me to test those transformations on a few different machines. Please take a look at the results below to gain better insight.

- Arduino Device
	- [Source code repo](https://github.com/onkoe/avr_f64_testing)
	- [Arduino MEGA 2560 R3 results](https://gist.github.com/onkoe/2d9936125812f917d2cbf5ec8e9c42c1)
- Other machines
	- [Shared source code](https://gist.github.com/onkoe/3cc2977cfc891711340da17ca23c9f49) for all below
	- MacBook Pro M1 on macOS and Asahi Linux (`aarch64`)
		-  Linux (`aarch64-unknown-linux-gnu`) [results](https://gist.github.com/onkoe/1eda11e69a75d6018a1057ab30bc3070)
		- macOS (`aarch64-apple-darwin`) [results](https://gist.github.com/onkoe/0de388b4be57ffad23b92e44e22169f8)
	- Ryzen desktop machine running Linux (`x86_64-unknown-linux-gnu`)[results](https://gist.github.com/onkoe/38f5ce12407f4118fd7f42900c03fad5)
	- Milk-V Duo (`riscv64gc-unknown-linux-musl`) [results](https://gist.github.com/onkoe/11313dbefb18418f916b65f6ac64381d)
	- I can do more if you'd like!

If there's any additional context I can give, or checking to be done, please let me know, and I'll happily look into it!

Also, I did check the `ufmt_float` crate to see if it was working properly! I assume that since the `f64 as u32` and ufmt_float'd `f64`are the same, it must be an issue with `f64` itself. Please see [the source](https://gist.github.com/onkoe/890253ce431e167fb4f9cfe219a48ead) and [serial output](https://gist.github.com/onkoe/703e76e64c42e84d93f2ef4be4d5dc50) of the original test to get more info.

Thank you for taking a look at my problem!

**Comment by apiraino** [comment]
@onkoe trying to figure out if this is a regression. I see that you tested on [nightly-2023-08-08](https://github.com/onkoe/avr_f64_testing/blob/main/rust-toolchain.toml), does it also reproduce on other versions of the rust compiler? Did you try using a stable compiler (if usable for your project)?

**Comment by Patryk27** [comment]
AVR doesn't support 64-bit numbers in Rust (operations on them should end up being linker errors, though 👀):
https://github.com/rust-lang/compiler-builtins/pull/527

The fundamental issue is that AVR intrinsics (e.g. operations for i32 or f32) require a special calling convention that hasn't been yet exposed from LLVM, preventing Rust's compiler-builtins from chiming in with its own implementation.

Some intrinsics (such as those for f32) get implemented by linking the resulting AVR binaries with [hand-written functions from GCC](https://github.com/gcc-mirror/gcc/blob/3eb9cae6d375d222787498b15ac87f383b3834fe/libgcc/config/avr/lib1funcs.S) (that's why working with AVR requires avr-gcc), but they don't provide implementations for f64.

tl;dr f64 is not supported, but (probably) should throw a linker error instead of trying to link with compiler-builtin's implementation (because that implementation uses a different calling convention than the LLVM's codegen expects)

**Comment by Patryk27** [comment]
Alright, got it! -- fix at: https://github.com/rust-lang/compiler-builtins/pull/558.

(note that merging that merge request in compiler-builtins doesn't yet solve the issue here, because then I'll need to prepare another merge request bumping compiler-builtins inside rustc - all should be ready within a few days)

As for the cause - since AVRs don't have a native support for floating-point operations, those must be implemented in software; usually implementations for things like that end up in a crate called compiler-builtins (linked above), which implements them in Rust using simple bitwise operations and whatnot.

There's a difference, though, between the ABI as expected by LLVM and the code that's present inside compiler-builtins - for "bigger" types (like i64, i128, f32 etc.), AVR requires a special calling convention (which defines things like "which register should contain which value") that is not yet supported by compiler-builtins (and not so easy to expose there).

So instead, we pull hand-written operations from GCC's standard library (which is why avr-gcc is required as the linker) - those definitions use correct registers - but the linking order is that stuff from compiler-builtins gets linked before GCC, that's why those "invalid" intrinsics have to be commented out inside compiler-builtins (so that the avr-gcc can use its own definitions).
